ast-checks/src/comment_finders.rs

```diff
@@ -0,0 +1,49 @@
+use crate::tree_utils::{next_siblings, prev_siblings, walk_kind};
+use rnix::{NodeOrToken, SyntaxElement, SyntaxKind::*, SyntaxNode};
+
+pub fn find_comment_within(node: &SyntaxNode) -> Option<String> {
+    // find the first TOKEN_COMMENT within
+    let tok = walk_kind(node, TOKEN_COMMENT).find_map(|item| match item {
+        NodeOrToken::Token(ref _tok) => Some(item),
+        NodeOrToken::Node(_) => {
+            unreachable!(
+                "`tok` should always be of kind==TOKEN_COMMENT,
+                which means that it's a SyntaxToken rather than
+                SyntaxNode"
+            );
+        }
+    })?;
+
+    let node_iter = next_siblings(&tok);
+    let doc = collect_comment_text(node_iter);
+    Some(doc).filter(|it| !it.is_empty())
+}
+
+pub fn find_comment_above(node: &SyntaxNode) -> Option<String> {
+    // note: the prev_siblings iterators includes self, which
+    // isn't a TOKEN_COMMENT, so we need to skip one
+    let node_iter = prev_siblings(&NodeOrToken::Node(node.clone())).skip(1);
+    let doc = collect_comment_text(node_iter);
+    Some(doc).filter(|it| !it.is_empty())
+}
+
+fn collect_comment_text(node_iter: impl Iterator<Item = SyntaxElement>) -> String {
+    // take all immediately subsequent TOKEN_COMMENT's text,
+    // skipping over whitspace-only tokens
+    // Note this would be more clearly written using map_while, but that
+    // doesn't seem to be on rust stable yet.
+    node_iter
+        .take_while(|node| match node {
+            NodeOrToken::Token(tok) if tok.kind() == TOKEN_COMMENT || tok.kind().is_trivia() => {
+                true
+            }
+            _ => false,
```

Comment:
So, what I meant here is that `tok.kind() == TOKEN_COMMENT || tok.kind().is_trivia()` _already_ evaluates to `true` or `false`, so you don't need the match here I think.

Can you pull the data from `NodeOrToken::Token` to perform the above _without_ a match?

Reply:
I think you can. You should do `.filter_map(|i| i.as_token()).take_while(|tok| tok.kind() == TOKEN_COMMENT || tok.kind().is_trivia())`

Reply:
https://docs.rs/rnix/0.8.0/rnix/enum.NodeOrToken.html#method.as_token

Reply:
Nice